### ISSUE TITLE
Check parameters of known messages in XaAES

### ISSUE THREAD

**Issue by th-otto** [freemint/freemint#291]

**Comment by th-otto** [comment]
I don't want to encourage lazy programmers to write new programs which are buggy ;) It's only a few lines to check whether Mxalloc() is available, and correctly use that.

As for the program flags: i'm not sure whether that helps much, since you have to set that flag somehow. Not much different from just making an entry in xaaes.cnf. I'm also not sure whether it is easy for XaAES to get access to the flags, since they are only part of the executable image, but not of the program in memory.

**Comment by mikrosk** [comment]
Ah ok, you meant it as an xaaes.cnf entry. Yep, that would work as well (although it would be limited, obviously, to XaAES only while PRG flags would work with N.AES, Geneva etc).

**Comment by mikrosk** [comment]
That PRG flags was meant mostly for advanced users, i.e. we'd provide a tool to edit the new flag and if they encounter a strange memory violation, we'd tell them "change this flag" instead of "boot without MP".

**Comment by th-otto** [comment]
> PRG flags would work with N.AES, Geneva etc).

Only if you change N.Aes / Geneva to also support that flag ;)

> if they encounter a strange memory violation, we'd tell them "change this flag"

We can also just tell them to add an entry to xaaes.cnf.

BTW, if we are really going to implement something like that, then we should start asking on AF which apps may be troublesome, so we can provide an xaaes.cnf with some default entries.

**Comment by mikrosk** [comment]
> Only if you change N.Aes / Geneva to also support that flag ;)

Hmm, actually you are right. My initial idea was to support this purely on kernel level but now as I think about it, XaAES has to tell the kernel that yes, this is the time to patch the Mxalloc() call (basically what we do now with TOS VDI + XaAES + MP).

So yes, xaaes.cnf now makes way more sense, esp. if we already have a flag array per app there.

> BTW, if we are really going to implement something like that, then we should start asking on AF which apps may be troublesome, so we can provide an xaaes.cnf with some default entries.

Yes, definitely. People use really various setups so everyone has his share of experience.

**Comment by skaftetryne** [comment]
> That PRG flags was meant mostly for advanced users, i.e. we'd provide a tool to edit the new flag and if they encounter a strange memory violation, we'd tell them "change this flag" instead of "boot without MP".

This feature already exists - set the "Global readable" flag for the offending application. Or maybe I misunderstand the purpose of this flag you suggests?

**Comment by th-otto** [comment]
Yes, setting the global readable flag would workaround that, but has some drawbacks:

- its not always obvious which is the offending application. Currently, the innocent receiver gets killed, not the buggy application
- every user has to do that again on his installation
- setting the global flag affects all memory allocations, not only the ones used for communication. If  a lot of your programs need that set, you might as well turn off memory protection completely

Unfortunately there was no response yet on AF

**Comment by skaftetryne** [comment]
There is no other practical way to work around the problem than changing the memory protection settings for the offending application. Instead of killing the recipient of the VA_START message, kill the sender and show a message telling the user to change the protection flag to "Global readable" for application X. "Global readable" is pretty safe I'd say, and *not* like disabling memory protection. Disabled memory protection allows any application to write to any other application's memory, "Global readable" does not.

I'm still not sure what the suggested flag or xaaes.cnf setting is supposed to do.

**Comment by th-otto** [comment]
Thats what this all about. Currently, the recipient gets killed by the kernel when trying to access the memory. But the kernel does not know that the memory was part of a VA_START protocol, nor does it know about the sender. And it also cannot simply display a message at that point.

**Comment by wongck68** [comment]
wow. this is a good idea.something like the  sscanf_s() in the new library with extra safety checks.
May be a wrapper around Appl_Write(), checking for memory type based on the message type.If the memory cannot be accessed by the recipient then just display a dialog about the issue and drop the call completely ( is there a reply to say it failed ? )
    On Monday, 20 February 2023 at 01:43:37 pm SGT, Thorsten Otto ***@***.***> wrote:  
 
 


Thats what this all about. Currently, the recipient gets killed by the kernel when trying to access the memory. But the kernel does not know that the memory was part of a VA_START protocol, nor does it know about the sender. And it also cannot simply display a message at that point.

**Comment by mikrosk** [comment]
@skaftetryne I think the main point is that right now all we have is what the **kernel** knows. @th-otto proposes to add more code to **xaaes** so we can get a wider playground.

Whether it will be a single setting in xaaes.cnf or a per-app flag, that's only an implementation detail.

**Comment by skaftetryne** [comment]
I understand what the purpose of this functionality is, but what I don't understand is the purpose of a per application flag or setting. Having such a flag/setting indicates that you know which application is problematic, in which case you either have to...

a) not use that application.
b) work around the problem by changing the memory protection flag for this application.
c) disable memory protection.

All of these options leaves a per application flag or setting useless.

But is is also very likely that I misunderstand this, could someone please explain exactly what the purpose of such a setting would be?

**Comment by mikrosk** [comment]
Yeah but if we distribute freemint+xaaes with the setting "for app XY, force the global readable flag when appropriate", the user doesn't have to do anything.

install freemint snapshot, download a ZIP file from a website, use, don't care about anything.

IMO this is a great enhancement against the current and pretty painful procedure (realise that it is the VA_START which is causing problems, realise who is the sender, find the PRGFLAGS tool/CPX, remember which flag to set, repeat every time you download an app).

**Comment by wongck68** [comment]
If the new function replaces the old Appl_write, 1) it calls the checking_function 2) then call Appl_write_old if it is OK 3) else put a dialog stating "Your API call violated memory access rules" and then just returns.
I am sure the user will tell the developer " hey it does not work" ... 
No need any flags. Isn't this what Thorsten 1st post ?
Sure the checking function will take up some time checking. You call will not be so fast.    On Wednesday, 22 February 2023 at 04:56:01 pm SGT, Miro Kropáček ***@***.***> wrote:  
 
 


Yeah but if we distribute freemint+xaaes with the setting "for app XY, force the global readable flag when appropriate", the user doesn't have to do anything.

install freemint snapshot, download a ZIP file from a website, use, don't care about anything.

IMO this is a great enhancement against the current and pretty painful procedure (realise that it is the VA_START which is causing problems, realise who is the sender, find the PRGFLAGS tool/CPX, remember which flag to set, repeat every time you download an app).

**Comment by skaftetryne** [comment]
> Yeah but if we distribute freemint+xaaes with the setting "for app XY, force the global readable flag when appropriate", the user doesn't have to do anything.

Ok, so it's not just about warning the user but modifying the memory protection flags when starting the application? If so then an appsetting makes sense, but then I don't understand the need to monitor messages. Can you change the protection status for already allocated memory?

From my understanding you and th-otto are suggesting two different approaches to the same problem. th-otto - monitor messages, alert the user when an application sends messages with pointers to private memory. mikrosk - add a feature to the xaaes.cnf appsettings to force allocated memory to be global for this particular application.

**Comment by wongck68** [comment]
ok understood.
I guess being able to stop the receiving application from being killed will be helpful.For me I normally try to change the prgflags  (fastload, st/ttram, global, private, super) to try to stop it crashing.
So all these time I was changing the wrong program 
    On Wednesday, 22 February 2023 at 06:56:41 pm SGT, skaftetryne ***@***.***> wrote:  
 
 



Yeah but if we distribute freemint+xaaes with the setting "for app XY, force the global readable flag when appropriate", the user doesn't have to do anything.


Ok, so it's not just about warning the user but modifying the memory protection flags when starting the application? If so then an appsetting makes sense, but then I don't understand the need to monitor messages. Can you change the protection status for already allocated memory?

From my understanding you and th-otto are suggesting two different approaches to the same problem. th-otto - monitor messages, alert the user when an application sends messages with pointers to private memory. mikrosk - add a feature to the xaaes.cnf appsettings to force allocated memory to be global for this particular application.

**Comment by skaftetryne** [comment]
The user must be notified. Just discarding the message is even worse than the current situation, because then things then just don't work and the user won't have any clue why not and probably blame MiNT because "it's working under MagiC".

If it's possible to change the protection flag of already allocated memory then I suggest the following:

Check each message sent to another application (remember that applications can send messages to themselves) for message types that we know includes pointers. Then check the protection flags of the memory address the pointer points to. If this is protected, change it to "Global readable". You will have to make all the application's memory readable, because you only know the start address of the data, not the size. Then pass the message on to the recipient. IMO you only need to do this check once for most application. If the first message containing a pointer points to globally readable/accessible memory you can safely assume that the program either use Mxalloc correctly, or that program flags has been correctly set.

There will be no need for any new program flags or application specific settings in xaaes.cnf in this case. It does not improve or solve anything that hasn't already been solved by the above.

Or as mikrosk suggests - make this an application setting in xaaes.cnf, where memory allocations for these applications is forced to "Global readable" if the program flags indicates a more restrictive default. In this case there is no need to scan messages. But OTOH you will need to either set this option for all apps, or maintain a list of applications. Then again - this is probably easier than covering all types of protocols.

Or even easier: Just improve the current alert that's presented to the user when an application is killed due to a memory violation. The current alert only makes sense to a developer. If the alert said something in the lines of "X was killed due to a memory protection violation, it attempted to access protected memory owned by Y. Please note that the problem may actually be caused by Y instructing X to do so, but failed to give X the proper access.". Not as slick, but should be a lot easier to implement.

**Comment by th-otto** [comment]
> The user must be notified. Just discarding the message is even worse than the current situation

Thats why i suggested to use per-application settings ;) Then the user would have a choice whether he wants to be notified or not (because he know that application is buggy)

>IMO you only need to do this check once for most application

I dont' think that such an optimization is possible. True, if an application doesn't use Mxalloc once, it won't use Mxalloc the next time either. But the memory block might be different.

>You will have to make all the application's memory readable, because you only know the start address of the data, not the size

That's why i wanted to look at the message, to prevent having to make *all* memory global readable. Memory protection can only be changed on a whole page, which is 8k in mint, that should be sufficient in most cases without knowing the actual size.

>it attempted to access protected memory owned by Y

To find the owner of the memory at the time when it is accessed, you would have to scan all processes. And as already mentioned, i think you can't display an alert at the time the sigfault happens. That could very badly interact with other memory protection uses, like outside. Some similar problem of running the critical alert error handler when getting BIOS errors in old AES.

**Comment by th-otto** [comment]
Maybe that's the case, too. But what i wanted to say, you just can't skip the test the 2nd time, you'll have to check it every time.

**Comment by xdelatour** [comment]
> And yes, i think that you can change the protection flags of already allocated memory. If not, then we'll need something like that anyway when implementing functions like mprotect(). But in this case, its not even required in the first place. It's good enough to verify that the memory is not readable by the receiving application, and then just discard the appl_write() call.

Mvalidate() and Maccess() are already available (syscall $141 and $17d) , $17e and $17f are reserved for mmap() and munmap(). It may be possible to use $17c for Mprotect(). To add to the ToDo list...

**Comment by mikrosk** [comment]
Btw, when looking into changelog between Thing 1.27 and 1.27b, I was surprised to find this improvement:

> Protection against crashes caused by programs such as ST-Guide when memory protection is active because they use private memory for the AV protocol

(in original _Schutz gegen Abstürze, die bei aktivem Speicherschutz durch Programme  wie ST-Guide verursacht werden, weil sie privaten Speicher für das AV- Protokoll einsetzen_)

To me it sounds like Thing did exactly what Thorsten has proposed in his first post.
